Desktop app: Authenticate with OAuth

psrpinto opened the pull request:
> This PR does not yet enable the OAuth flow in the desktop app, the regular flow is still in use. The OAuth flow introduced here will be enabled in an upcoming PR.

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Requires 172504-ghe-Automattic/wpcom (already deployed)

Related to  #89736 #72754 #55611


## Proposed Changes

This PR makes it so that logging into the Desktop app happens in the user's external browser, instead of directly in the app. So, instead of the Desktop app showing calypso's login screen, it shows just a button that when clicked opens calypso's login screen in the user's browser.

Once the user has completed login in their browser, they are redirected to the desktop app, through a URL that carries the `access_token`, e.g.:

```
wpdesktop://token#access_token=foo
```

The access token is then "traded" for a cookie by submitting the login form.

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

Currently, the desktop app is unusable for users who, for example, have passkeys. There are also other cases where login in the desktop app is broken, see linked issues above. Doing the login externally instead of directly in the desktop app would fix all authentication-related issues.

Prior analysis on this has been done in p1717608504988349-slack-C029GN3KD and pdKhl6-49R-p2. 

## Screen captures

https://github.com/user-attachments/assets/a10585a1-b8b4-4292-88b5-8f6fdd333498

<img width="801" alt="Screenshot 2025-02-07 at 13 01 43" src="https://github.com/user-attachments/assets/206cc3ec-071b-4333-9f78-5578313eb30e" />

<img width="801" alt="Screenshot 2025-02-07 at 13 02 51" src="https://github.com/user-attachments/assets/68a8cc0b-ce40-4bde-a9a2-1a3a1aef582e" />

## Proposed changes

- [x] Add a new route and view to calypso that displays a new login screen (only present on the desktop app)
- [x] Add a new route and view to calypso that finalizes login in the desktop app, after the user authenticated in their browser
- [x] Make it so that the desktop app handles incoming links to `wpdesktop://` ([electron protocol handling](https://www.electronjs.org/docs/latest/tutorial/launch-app-from-url-in-another-app))
- [x] Have the desktop app intercept `wpdesktop://token` URLs, and call a new `/log-in/desktop/finalize` route in calypso

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

The changes in this PR are not in use yet (the desktop app is still using the previous login flow), so no testing should be required.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

psrpinto committed:
Register app as a protocol handler

psrpinto committed:
Start authentication flow in the external browser

psrpinto committed:
Remove unnecessary module

isSingleInstance() was only called in one place (so we inlined it), and anotherInstanceStarted() was unused.

psrpinto committed:
Move single instance check to as early as possible

And exit immediately when not the first instance.

psrpinto committed:
Handle incoming URLs

psrpinto committed:
Add routes for desktop login

psrpinto committed:
Implement desktop login page

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/17234541

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Hi @psrpinto, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include all of the following strings:
- `We were not able to log you in. Please try again.`
- `Authorize with WordPress.com to get started`

Thank you in advance!

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/17234541).

jsnajdr commented on client/login/controller.js:
```diff
@@ -135,6 +136,19 @@ export async function login( context, next ) {
+		<DesktopLogin action="finalize" error={ hash?.error } accessToken={ hash?.access_token } />
+	);
+	return next();
+}
```
The two new route handlers don't need to be async functions: there is no `await` inside them. They pass control by calling the `next()` function, and the return value (promise) from the handler is not used anywhere.

`return next()` can be replaced with `next()`: the function will return anyway on the next line.

psrpinto replied:
Good points, fixed.

jsnajdr commented on client/login/desktop-login/index.tsx:
```diff
@@ -0,0 +1,115 @@
+const loginUrl = '/desktop-start-login';
+
+type Props =
+	| { action: 'start'; accessToken?: never; error?: never }
```
Isn't it more correct when the `start` type is just `{ action: 'start' }`? Then you'll to reorganize the `DesktopLogin` code to guard access to `accessToken` and `error` props with `isStart` and `isFinalize` condition. And that looks like an improvement that clarifies the code.

psrpinto replied:
I agree with your assessment. However, since the component has now been split in two, this code is no longer present.

jsnajdr commented on client/login/desktop-login/index.tsx:
```diff
@@ -0,0 +1,115 @@
+ * We then use that access token to submit the login form, which will set the cookie,
+ * and thus log the user in.
+ */
+export default function DesktopLogin( props: Props ) {
```
Do the `start` and `finalize` versions of the component have anything in common? Except the `<Main>` wrapper and the error notice?

Seems to me we'd be better off with two components, `DesktopLoginStart` and `DesktopLoginFinalize`. Currently we're merging two components together, with a lot of conditional logic inside.

psrpinto replied:
This makes sense, I'm working on splitting it into two components.

psrpinto replied:
I've pushed a commit that splits it into two components.

jsnajdr reviewed (commented):
The PR description mentions this URL?

> ```
> wpdesktop://token#access_token=foo&expires_in=1209600
> ```

My question is whether the `expires_in` param is really there? Where it's coming from? Our OAuth token typically never expire.

psrpinto committed:
Functions don't need to be async

psrpinto committed:
No need to return the result of next()

psrpinto commented:
> >  wpdesktop://token#access_token=foo&expires_in=1209600
>
> My question is whether the expires_in param is really there? Where it's coming from? Our OAuth token typically never expire.

The `expires_in` parameter is in the `Location` header of the response after the user authorizes at https://public-api.wordpress.com/oauth2/authorize?response_type=token&client_id=41936&redirect_uri=wpdesktop%3A%2F%2Ftoken&scope=global

```
wpdesktop://token#access_token=redacted&expires_in=1209600&token_type=bearer&site_id=0&scope=global
```

Indeed, we only use the `token`, we don't use the `expires_in` nor other parameters, I just included it in the PR description to illustrate the form of the URL. I now edited the PR description to only keep the `token` in the example URL.

psrpinto committed:
Split desktop login views into two components

psrpinto committed:
Do not yet use the new desktop login flow

psrpinto committed:
Merge branch 'trunk' into desktop-oauth

psrpinto commented:
@jsnajdr I've now addressed all your remarks (thanks!), this should be ready for re-review.

I've also pushed https://github.com/Automattic/wp-calypso/pull/98345/commits/1000d1a5093081b0244727a6a873bac92b3f9d61, which disables the new login flow in the desktop app, for now. I did this so that I can deploy the calypso changes first, since I would like to be able to test the new desktop login flow with the production calypso.  This is necessary because the non-dev build of the desktop app mounts the production calypso so the calypso changes introduced here need to be deployed first.

This means merging this PR should be low-risk, as the desktop app will keep using the previous login flow. I will then open a new PR where all the testing can happen.

jsnajdr commented on client/login/desktop-login/start.tsx:
```diff
@@ -0,0 +1,48 @@
+					<Notice
+						status="is-error"
+						onDismissClick={ () => setError( undefined ) }
+						translate={ translate }
```
You don't need to pass `translate` here, the `Notice` component gets it from the `localize()` HOC.

psrpinto replied:
Thanks, fixed. I was importing the non-localize-wrapped component:

```ts
import { Notice } from 'calypso/components/notice'
```

I now changed it to the default export:

```ts
import Notice from 'calypso/components/notice'
```

And we're no longer passing the `translate` prop.

jsnajdr reviewed (approved):
Looks good :+1:

jsnajdr commented on client/login/desktop-login/start.tsx:
```diff
@@ -0,0 +1,48 @@
+import { useState } from 'react';
+import FormattedHeader from 'calypso/components/formatted-header';
+import Main from 'calypso/components/main';
+import { Notice } from 'calypso/components/notice';
```
Related to the `translate` prop: if you import the default import from the same file:
```js
import Notice from 'calypso/components/notice';
```
then that will be the version wrapped with `localize`. That's insidious, I'm going to fix that and export only one version.

psrpinto replied:
Thanks for opening https://github.com/Automattic/wp-calypso/pull/99637 to fix this, it was indeed confusing to me.

psrpinto committed:
No need pass translate as a prop to Notice
